### ISSUE TITLE
AutoMapper tests are failing when using FEC

### ISSUE THREAD

**Issue by dadhi** [dadhi/FastExpressionCompiler#196]
AutoMapper issue https://github.com/AutoMapper/AutoMapper/pull/3103#issuecomment-497096846

**Comment by jbogard** [comment]
I found the issue here. Here's a failing expression:

```c#
var expression = Lambda<Func<Source, Dest, ResolutionContext, Dest>>(
    Block(
        Condition(
            Equal(srcParam, Constant(null)),
            Default(typeof(Dest)),
            Block(typeof(Dest), new[] {typeMapDestParam},
                Assign(typeMapDestParam, Coalesce(destParam, New(typeof(Dest).GetConstructors()[0]))),
                TryCatch(
                    /* Assign src.Value */
                    Block(resolvedValueParam,
                        Block(
                            Assign(resolvedValueParam,
                                Condition(Or(Equal(srcParam, Constant(null)), Constant(false)),
                                    Default(typeof(int)),
                                    Property(srcParam, "Value"))
                            ),
                            Assign(Property(typeMapDestParam, "Value"), resolvedValueParam)
                        )
                    ),
                    Catch(exceptionParameter, Block(
                        Throw(New(constructorInfo, 
                            Constant("Error mapping types."), 
                            exceptionParameter,
                            Constant(typeMap.Types),
                            Constant(typeMap),
                            Constant(memberMap)
                            )
                        ),
                        Default(typeof(int))
                    ))
                ),
                typeMapDestParam
            )
        )
    ),
    srcParam,
    destParam,
    Parameter(typeof(ResolutionContext), "ctxt")
);
```

Note that my `catch` block has a "return" value for the `TryCatch` expression. This is because both sides of the try/catch have to return the same type, and we cheated by having the "catch" block "return" a value of `default(int)`.

The fix is to make the `try` block have a return type of `void`:

```c#
var expression = Lambda<Func<Source, Dest, ResolutionContext, Dest>>(
    Block(
        Condition(
            Equal(srcParam, Constant(null)),
            Default(typeof(Dest)),
            Block(typeof(Dest), new[] {typeMapDestParam},
                Assign(typeMapDestParam, Coalesce(destParam, New(typeof(Dest).GetConstructors()[0]))),
                TryCatch(
                    /* Assign src.Value */
                    Block(typeof(void), new[] {resolvedValueParam},
                        Block(
                            Assign(resolvedValueParam,
                                Condition(Or(Equal(srcParam, Constant(null)), Constant(false)),
                                    Default(typeof(int)),
                                    Property(srcParam, "Value"))
                            ),
                            Assign(Property(typeMapDestParam, "Value"), resolvedValueParam)
                        )
                    ),
                    Catch(exceptionParameter, 
                        Throw(New(constructorInfo, 
                            Constant("Error mapping types."), 
                            exceptionParameter,
                            Constant(typeMap.Types),
                            Constant(typeMap),
                            Constant(memberMap)
                            )
                        )
                    )
                ),
                typeMapDestParam
            )
        )
    ),
    srcParam,
    destParam,
    Parameter(typeof(ResolutionContext), "ctxt")
);
```

Which makes more sense, but the C# expression compiler handled this scenario fine.

Relevant StackOverflow question: https://stackoverflow.com/questions/14365122/expression-convert-within-expression-trycatch

**Comment by dadhi** [comment]
@jbogard Thanks for pointing it out! I am slowly moving towards the fix optimizing some things on the way.

**Comment by dadhi** [comment]
fixed. at least this case.

**Comment by dadhi** [comment]
There is still a lot to fix in #222

**Comment by dadhi** [comment]
closing, the remaining work is in #222
